Add missing branch_remote_name

Finii opened the pull request:
## Step 1: Add proper Error object

**[why]**  
The libgit2 has a lot of error codes that we migth want to transport.    
At the moment that is not possible and we can just generate some random    
text messages that are hard to decifer for our users.    
    
**[how]**  
Use `std::system_error` which is able to transport error codes.    
    
Define the error categories for the git related error codes, so that the    
standard library can convert the number to text.    
    
Enable the `test_Error.cc` and put something meaningful into it.    
    
Add some extra constructors to the Error object that just take a string.    
They were used before when a `std::exception` has been utilized which can    
not hold error codes. The extra constructor uses in this case the error    
code `GIT_EUSER` as reserved exactly for this purpose by `libgit2`.    

## Step 2: Implement missing function

**[why]**    
The function `branch_remote_name()` is already used but never defined.    
This results in a library with missing symbols and when the function    
depending on that symbol is called a crash.    
    
**[how]**    
There already was an almost complete but commented out implementation.    
Add evaluation of the error code to that implementation and use the    
Error object to raise an exception with code and explanantion.  

## Step 3 to n

Still missing are tests for the just implemented function.  
But that is true for more wrapper functions.

Also the other wrapper functions do not transport the gitlib2 error code and they do not raise Error on errors, but the error handling is somehow rudimentary implemented by return values. That should be fixed.

Finii committed:
Add proper Error class and enable Error tests

[why]
The libgit2 has a lot of error codes that we migth want to transport.
At the moment that is not possible and we can just generate some random
text messages that are hard to decifer for our users.

[how]
Use std::system_error which is able to transport error codes.

Define the error categories for the git related error codes, so that the
standard library can convert the number to text.

Enable the test_Error.cc and put something meaningful into it.

Add some extra constructors to the Error object that just take a string.
They were used before when a std::exception has been untilized which can
not hold error codes. The extra constructor uses in this case the error
code GIT_EUSER as reserved exactly for this purpose by libgit2.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
wrapper: Implement missing branch_remote_name()

[why]
The function branch_remote_name() is already used but never defined.
This results in a library with missing symbols and when the function
depending on that symbol is called a crash.

[how]
There already was an almost complete but commented out implementation.
Add evaluation of the error code to that implementation and use the
Error object to raise an exception with code and explanantion.

Fixes: #19

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
Forgot to add the new file `Error.cc` :grimacing: just added the changes. Force push.  
Hmm, there is still a (CI) problem with an invalid free, examining...

...

Hmm, inside the libgit2 library?

```
#4  0x00007ffff767d3d9 in __interceptor_free (ptr=0x555555795920) at ../../../../src/libsanitizer/asan/asan_malloc_linux.cc:127
#5  0x00007ffff6eb3d62 in git_buf_dispose () from /lib/x86_64-linux-gnu/libgit2.so.28
#6  0x00007ffff6eb28fa in git_branch_remote_name () from /lib/x86_64-linux-gnu/libgit2.so.28
#7  0x00007ffff7561309 in git::branch_remote_name (repo=0x612000004b40, branch_name="master") at ../src/wrapper_functions.cc:178
#8  0x0000555555594304 in ____C_A_T_C_H____T_E_S_T____0 () at ../tests/test_GitRepository.cc:100
```

Finii commented:
Fixed by calling the constructor of the `git_buf` :grimacing: 

Also added code to free the buffer in case of a failure, as the docu does not say if or if not the buffer might get filled in that case.

alt-graph commented on include/libgit4cpp/Error.h:
```diff
@@ -25,36 +25,114 @@
  * }
  * catch (const git::Error& e)
  * {
  *     std::cerr << e.what() << "\n";
```
Maybe add an example for using the error code?

alt-graph replied:
... I mean, for using `.code()`.

alt-graph commented on include/libgit4cpp/Error.h:
```diff
@@ -25,36 +25,114 @@
+    Error(int ev)
+        : std::system_error(ev, git_category())
+        { }
+    Error(int ev, const std::string& what)
```
The order of parameters matches `std::system_error`, so this is probably the best way to go. But just to mention it: In `doocs::`, `hlc::` and probably other contexts we have `ErrorCode` classes that take the parameters in the inverse order, and those `ErrorCode` classes are derived from `Error` classes that just contain a string. But I would leave it as implemented here.

Finii replied:
> In doocs::, hlc:: and probably other contexts we have ErrorCode classes that take the parameters in the inverse order

:sob: Why would one write `ErrorCode` that does not inherit from `std::system_error`.
Probably at least in the longer run, maybe that other libs should be fixed to use a "standard conforming" error code object.

alt-graph commented on include/libgit4cpp/Error.h:
```diff
@@ -25,36 +25,114 @@
 
 namespace git {
 
+class git_category_impl : public std::error_category {
```
`<nitpick>`Should be GitCategoryImpl according to the style guide.`</nitpick>`

Finii replied:
But then again "our" style guide clashes with standard library customary style.
I can change that, as it is just internal.

We also had this with `SomeThing::value_type` which must be `SomeThing::ValueType` according to the style guide which is against any expectations C++ programmers will have. 

Same here, people would expect the repository type to be `git_category` because that is the form it has everywhere else, and its implementation , well, is named not `git_category_impl`, but ... `git_categoryImpl` maybe?

Luckily the guide allows to deviate when there are reasons.

Finii replied:
(Just rediscovered the comment above, pending, in an old browser tab. Wanted to reboot and checked all 500 open tabs :grimacing:)

alt-graph replied:
I have said my piece. But as this is internal... :shrug:

But now that I look at this: Maybe the class should not be at `namespace git` scope, but hidden even better. It could probably be an implementation detail of the `git_category()` function, couldn't it?

Finii replied:
I put `git_category_impl` into namespace `git::detail`.

alt-graph commented on include/libgit4cpp/Error.h:
```diff
@@ -25,36 +25,114 @@
+        GIT_EINDEXDIRTY     = -34, ///< Unsaved changes in the index would be overwritten
+        GIT_EAPPLYFAIL      = -35, ///< Patch application failed
+        GIT_EOWNER          = -36, ///< The object is not owned by the current user
+        GIT_TIMEOUT         = -37  ///< The operation timed out
```
Are we reasonably sure that these enumerators are not going to change in libgit2 under the hood? And doesn't the library provide enums or at least constants for them?

Last point: enumerators should be lower_case according to the style guide.

Finii replied:
The case is just identical to the case of the errors mentioned in the libgit2 documentation.
I specifically kept the naming to make it easier for users.
Although THAT of course are macros, and C.

I believe you cling to much to the style guide. I think it is ok for server programmers and mainly for our own libraries, but in some places it not only allows to do it differently, we **should** implement it in the most sane way, and that is not always what the style guide foresees for brand new code without background.

alt-graph replied:
There are a few things I need to clarify here:
- First, when I wrote my comment in January, we actually had no rule on enumerators in the style guide. I thought there was one, but I thought wrong. I have added a proposal for an "enumerators_use_snake_case" rule a few days ago. We can discuss its merit there.
- Second, since there is no agreed rule, I withdraw this comment.

But in general, I still believe that the style guide defines useful default rules for how to name things, and that it is very helpful for everyone involved to follow these rules. Sure, one may choose to deviate from them, but then one should explain why.

Finii replied:
Agree

Finii replied:
Removed the enum completely :tada:

alt-graph commented on src/Error.cc:
```diff
@@ -0,0 +1,73 @@
+/**
+ * \file   Error.cc
+ * \author Lars Fröhlich, Jannik Wöhnert
```
Nah. ;)

Finii replied:
I believe we should drop all those `\author` lines anyhow; I hate them.

I also hate the copyright-year line too btw, but to a lesser extend. 

Can we not say something like `\author The project authors`or something? :grimacing:

alt-graph replied:
I honestly have no clue. I think these author lists have some legal meaning to point out which individuals worked on the code at which point in time (to figure out if their copyright is still valid at some point in the future). So in theory this should probably look like this:
```
\copyright
1999 D. Duck
2000-2012, 2015 M. Mouse
2020 Tick, Trick, Track
```
But... :shrug:

alt-graph replied:
... in any case, I did not write a line of this. :-D

alt-graph commented on tests/test_Error.cc:
```diff
@@ -29,44 +28,30 @@
-    REQUIRE(e.what() == "Test"s);
+    // Default error code is -7 == GIT_EUSER
+    git::Error e("Test");
+    REQUIRE(e.what() == "Test: GIT_EUSER"s);
```
Interesting... I wouldn't have expected this. Is this format guaranteed by the standard? In any case, no objection as long as it does not fail on any of our platforms.

Finii replied:
This is rather old. I am not sure it is guaranteed by the standard, but if it is implemented this way in "our 4 major compilers" I think it is save enough to test.
What is guaranteed is that the string parts are there, we could instead check if all parts exist.

alt-graph replied:
Well, as long as it works... 🤷

alt-graph commented on src/wrapper_functions.cc:
```diff
@@ -172,17 +172,17 @@ LibGitReference branch_lookup(git_repository* repo, const std::string& branch_na
+    git_buf buf{ };
+    auto result = git_branch_remote_name(&buf, repo, branch_name.c_str());
+    if (result) {
+        git_buf_dispose(&buf);
```
It might be better to put the `git_buf_dispose()` into a `gul14::finally()` cleanup object.

Finii replied:
Right.

Finii replied:
> Also added code to free the buffer in case of a failure, as the docu does not say if or if not the buffer might get filled in that case.

Checked the code, `git_branch_remote_name` is in fact macro magick :-(
But we indeed need to free it always.

Implemented via finalizer.

alt-graph reviewed (changes requested):
Apart from some mismatch with the style guide, this looks excellent. I am a bit afraid of duplicating the git-internal error codes, but if there is really no other way...

alt-graph commented on include/libgit4cpp/Error.h:
```diff
@@ -25,36 +25,114 @@
 public:
-    using std::runtime_error::runtime_error;
+    using std::system_error::system_error;
+    Error(int ev)
```
Maybe `ev` could be `error_code` or something like that. What does `ev` stand for?

Finii replied:
It can not be `error_code` as it is not an error code but an int :grimacing: 

I _guess_ it means 'error value' as to distinguish it from a error_code object.

Just mimicked the naming from 'upstream':

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/5cad2b13-ebed-4199-8f99-11d9ea9c28bb)

alt-graph replied:
They are funny on cppreference: "Constructs with underlying *error code* `ev`"... Which just proves that naming is hard. :-D

alt-graph reviewed (approved):
I think this is largely good to merge. Some comments are still open, maybe some improvements can still be made.

Finii commented:
Rebase on main, force push
Now addressing things

Finii committed:
Error: Put git_category_impl into detail namespace

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
Error: Correct authorship

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
wrapper_functions: Simplify branch_remote_name() cleanup

[why]
The libgit2 buffer needs to be cleaned up in the end always. At the
moment we handle that manually.

[how]
Use a finalizer object that cleans up the allocated buffer when the
function ends in all cases.

It is always safe to call, because it will not free if it has not yet
been assigned/allocated.

[note]
I'm not sure about the variable name _ (underscore). It is in wide use
but in other places in libgit4cpp we call it 'cleanup'.
For me just an underscore is more obvious that it will not be used for
anything but RAII.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
Error: Implement missing git_category ctor

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
Error: doc: Add example code for error category

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
Error: Redesign git_category without enum duplication

Also fix the implicit conversion and some example code.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
tests: Error: Fix test tag

[why]
Usually we have the test tag being the same name as the test file name.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
Sigh. Directly relying on the values of the library is hard...

| enum entry | since |
|---|---|
| GIT_EINDEXDIRTY | 0.28.0 |
| GIT_EAPPLYFAIL | 0.28.0 |
| GIT_EOWNER | 1.5.0 |
| GIT_TIMEOUT | 1.7.0 |
| GIT_EUNCHANGED | >1.7.2 |

Lets see \*flexes fingers\*

Finii committed:
Error: Allow building with any libgit2 version

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
This has some explanations of the error categories etc (5 part series)
http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-1.html

alt-graph commented on src/wrapper_functions.cc:
```diff
@@ -169,17 +170,16 @@ LibGitReference branch_lookup(git_repository* repo, const std::string& branch_na
+    auto _ = gul14::finally([buf_addr = &buf]() { git_buf_dispose(buf_addr); });
+    auto result = git_branch_remote_name(&buf, repo, branch_name.c_str());
+    if (result) {
+        throw Error{ result, gul14::cat("branch_remote_name: ", git_error_last()->message) };
```
I would find `error` more readable than `result` here, just because of the `if (error)`.

Finii replied:
Done.

While looking on this code I realized that `LibGitBuf` is probably useless (in `types.h`) - I tried to use it...

alt-graph reviewed (approved):
Looks good, I especially like the new example.

Left one really tiny comment. Feel free to merge with or without addressing it.

Finii committed:
wrapper_functions: Improve variable name

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
Error: Show implicit error_code conversion in example

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
> Looks good, I especially like the new example.

The real magic part was missing, the implicit conversion to `error_code`; added that also to the example (and test).
